### PR TITLE
[2.0] [BACKPORT] of [1.x][FEATURE] CUDA graphs support (#19142)

### PR DESCRIPTION
**Pull request by DickJC123**
This PR ports the CUDA Graphs support (by default off, enabled by MXNET_ENABLE_CUDA_GRAPHS=1) to 2.0.  For gluon models with static_shape and static_alloc both true, enabling CUDA Graphs can significantly reduce CPU overheads, leading to end-to-end perf improvements.  See PR https://github.com/apache/incubator-mxnet/pull/19142 for additional details.

### CODE REVIEW

**Comment by chinakook** [comment]
@DickJC123 Hi, guru. I've tried to test your commit on my faster rcnn, and it's great. It really get 17% performance increase on inferencing, but It can't be work with multi-thread such as one model shared with multiple threads. It will eat all the GPU memory gradually.

**Comment by leezu** [comment]
For 2.0, should cuda graphs be enabled by default?

Is there any update on "I'm still exploring the ways of automatically testing newly added operators in order for the feature to be able to be on by default, but I do not consider this the scope of this PR, as v1.x branch is not really supposed to get many more operators (I will do that in the PR to master). Generally this would involve testing operators with hybridize(static_alloc=True, static_shape=True) (which generally should be tested much more as right now testing of this functionality is really limited, even though it is widely used)." @ptrendx https://github.com/apache/incubator-mxnet/pull/19142#discussion_r488793102?

**Comment by chinakook** [comment]
> @DickJC123 Hi, guru. I've tried to test your commit on my faster rcnn, and it's great. It really get 17% performance increase on inferencing, but It can't be work with multi-thread such as one model shared with multiple threads. It will eat all the GPU memory gradually.

I've tried compile mxnet with cuda 11.6 and the latest cudnn version. There are no bugs with cuda graphs.

File: src/imperative/cuda_graphs.h [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  // Is the Op OK to make part of a CUDA Graph?
+  bool OpOK(const std::shared_ptr<exec::OpExecutor>& exec) {
+    static auto& fgraphcompatible = Op::GetAttr<FIsCUDAGraphsCompatible>("FIsCUDAGraphsCompatible");
+    static auto& fstateful        = Op::GetAttr<FCreateOpState>("FCreateOpState");
```
Comment (ptrendx) [inline src/imperative/cuda_graphs.h:514]: It is redundant once you check for the stateful compute.

File: src/operator/contrib/count_sketch.cu [hunk 2]
```diff
@@ -93,6 +93,9 @@ __global__ void sketch_backward_kernel(const int nthreads,
   // only calculate gradient regarding x
   // can also calculate gradient regarding s if needed
   const int index    = blockIdx.x * blockDim.x + threadIdx.x;
+  if (index >= nthreads) {
```
Comment (ptrendx) [inline src/operator/contrib/count_sketch.cu:96]: That test was breaking the CI for a looong time, thanks for the fix :-). Could you search for the issues about it and close them once this is merged?

Comment (DickJC123) [reply src/operator/contrib/count_sketch.cu:96]: This was a 'tack on' commit that was also in an already merged PR https://github.com/apache/incubator-mxnet/pull/20876, so it's part of master already. Per your suggestion here, I've now closed the open issue mentioning test_countsketch flakyness: https://github.com/apache/incubator-mxnet/issues/10988

File: src/imperative/cuda_graphs.h (hunk 1 above)
Comment (DickJC123) [reply src/imperative/cuda_graphs.h:514]: I'll remove the redundant check involving `fstateful`.

File: src/operator/tensor/elemwise_unary_op_basic.cu [hunk 3]
```diff
@@ -115,7 +115,10 @@ void ShapeComputeGPU(const nnvm::NodeAttrs& attrs,
                   mshadow::Stream<gpu>::GetStream(s));
 }
 
-NNVM_REGISTER_OP(shape_array).set_attr<FCompute>("FCompute<gpu>", ShapeComputeGPU);
+NNVM_REGISTER_OP(shape_array)
```
[the remark this answers is not in the document]
Comment (DickJC123) [reply src/operator/tensor/elemwise_unary_op_basic.cu:118]: I'm not sure whether I saw a failure, or whether I marked it incompatible on general principles.  But basically, this op does a cpu->gpu copy: https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/elemwise_unary_op_basic.cu#L111-L115

Now for CUDA Graphs to function, we rely on the fact that the tensor's gpu dptr will remain unchanged, but what about  the TBlob instance?  ShapeComputeGPU() is receiving a reference to the inputs and is not copying the TBlob, but can the op rely on the fact that the TBlob is not copied at some higher level of graph handling?  A copy of the input TBlob instance would preserve the dptr, but would change the cpu-memory pointer value returned by in_data.shape_.data().

So bottom line, I think this might work to have the Op declared 'CUDA Graphs compatible', but it relies on a different 'contract' with the rest of the software.

**Comment by DickJC123** [comment]
Merging now based on the approval from @ptrendx, our successful experience with CUDA Graphs on the 1.x branch, and the notion that the CUDA Graphs is off by default (enabled via MXNET_ENABLE_CUDA_GRAPHS=1).
